### PR TITLE
Promisify azure-iot-device

### PR DESCRIPTION
**Pull request by kpajdzik**

### CODE REVIEW

File: common/core/src/promise_utils.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+'use strict';
+
+export type MultiValueCallback<TResult1, TResult2> = (result1?: TResult1, result2?: TResult2) => void;
+export type NoErrorCallback<TResult> = (result?: TResult) => void;
+export type ErrorCallback = (error?: Error) => void;
+export type Callback<TResult> = MultiValueCallback<Error, TResult>;
+
+export function callbackToPromise<TResult>(callBackOperation: (callback: Callback<TResult>) => void): Promise<TResult> {
+  return new Promise((resolve, reject) => {
```
Comment (amarzavery) [inline common/core/src/promise_utils.ts]: `return new Promise<TResult>`

**Review by amarzavery — APPROVED** [review]

File: device/core/src/x509_authentication_provider.ts [hunk 2]
```diff
@@ -27,11 +27,19 @@ export class X509AuthenticationProvider implements AuthenticationProvider {
    *
    * @param callback function that will be called with either an error or a set of device credentials that can be used to authenticate with the IoT hub.
    */
-  getDeviceCredentials(callback: (err: Error, credentials?: TransportConfig) => void): void {
+  _getDeviceCredentials(callback: (err: Error, credentials?: TransportConfig) => void): void {
```
Comment (kpajdzik) [inline device/core/src/x509_authentication_provider.ts]: Change type here

File: provisioning/device/src/polling_state_machine.ts [hunk 3]
```diff
@@ -237,20 +237,44 @@ export class  PollingStateMachine extends EventEmitter {
     });
   }
 
-  register(request: RegistrationRequest, callback: (err?: Error, result?: DeviceRegistrationResult, response?: any) => void): void {
+  _register(request: RegistrationRequest, callback?: Callback<DeviceRegistrationResult>): void {
```
Comment (kpajdzik) [inline provisioning/device/src/polling_state_machine.ts]: This transformation is incorrect

File: device/core/src/device_client.ts [hunk 4]
```diff
@@ -96,21 +96,41 @@ export class Client extends InternalClient {
    *
    * *Note: After calling this method the Client object cannot be reused.*
    *
-   * @param closeCallback Function to call once the transport is disconnected and the client closed.
+   * @param {Callback<results.Disconnected>} closeCallback Function to call once the transport is disconnected and the client closed.
```
Comment (kpajdzik) [inline device/core/src/device_client.ts]: Here's sample documentation. I didn't add it to rest of the method as I want to hear feedback first.

File: device/core/src/module_client.ts [hunk 5]
```diff
@@ -182,16 +191,36 @@ export class ModuleClient extends InternalClient {
     this._methodClient.invokeMethod(deviceId, actualModuleId, actualMethodParams as MethodParams, actualCallback);
   }
 
+  invokeMethod(deviceId: string, methodParams: MethodParams, callback: Callback<MethodResult>): void;
+  invokeMethod(deviceId: string, moduleId: string, methodParams: MethodParams, callback: Callback<MethodResult>): void;
+  invokeMethod(deviceId: string, moduleIdOrMethodParams: string | MethodParams, methodParamsOrCallback: MethodParams | Callback<MethodResult>, callback?: Callback<MethodResult>): Promise<MethodResult> | void {
+    if (callback) {
+      return this._invokeMethod(deviceId, moduleIdOrMethodParams as string, methodParamsOrCallback as MethodParams, callback);
+    } else if ((methodParamsOrCallback instanceof Function)) {
+      return this._invokeMethod(deviceId, moduleIdOrMethodParams as MethodParams, methodParamsOrCallback as Callback<MethodResult>);
```
Comment (kpajdzik) [inline device/core/src/module_client.ts:203]: Can be simplified to one if.

File: common/core/src/promise_utils.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ */
+export type Callback<TResult> = DoubleValueCallback<Error, TResult>;
+
+/**
+ * Defines type describing regular callback with three results - two are the result values, the other one is Error.
+ *
+ * @template TResult1 - Type of the first result value.
+ * @template TResult2 - Type of the second result value.
+ */
+export type TripleValueCallback<TResult1, TResult2> = (error?: Error, result1?: TResult1, result2?: TResult2) => void;
```
Comment (kpajdzik) [inline common/core/src/promise_utils.ts:38]: I'm not a big fan of the name so if someone has better ideas I'd appreciate it. Same for `DoubleValueCallback`.

**Review by ghost — APPROVED** [review]

File: service/src/amqp.ts [hunk 7]
```diff
@@ -413,34 +415,40 @@ export class Amqp extends EventEmitter implements Client.Transport {
...
-        done(translateError('AMQP Transport: Could not connect', err));
-      } else {
-        done(null, new results.Connected());
-      }
-    });
+  connect(done?: TripleValueCallback<results.Connected, IncomingMessage>): Promise<ResultWithIncomingMessage<results.Connected>> | void {
+    return tripleValueCallbackToPromise((_callback) => {
+      this._fsm.handle('connect', (err) => {
+        if (err) {
+          done(translateError('AMQP Transport: Could not connect', err));
```
Comment (kpajdzik) [inline service/src/amqp.ts:426]: TripleValueCallback not necessary here

**Review by amarzavery — APPROVED** [review]

**Pull request closed without merging**